[VPlan] Implement VPWidenLoad/StoreEVLRecipe::computeCost().

ElvisWang123 opened the pull request:
Currently the EVL recipes transfer the tail masking to the EVL.
But in the legacy cost model, the mask exist and will calculate the
instruction cost of the mask.
To fix the difference between the VPlan-based cost model and the legacy
cost model, we always calculate the instruction cost of the mask in the EVL recipes.

Note that we should remove the mask cost in the EVL recipes when we
don't need to compare to the legacy cost model.

This patch also fixes #109468.

ElvisWang123 committed:
[LV][VPlan] Not adding shuffle cost when store loop invariant value.

This patch makes VPWidenMemoryRecipe and the legacy model not adding
reverse shuffle cost when the stored value is loop invariant.

This patch also fixes #109468.

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/vectorize-force-tail-with-evl-uniform-store.ll:
```diff
@@ -0,0 +1,160 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt < %s --prefer-predicate-over-epilogue=predicate-dont-vectorize --passes=loop-vectorize -mcpu=sifive-p470 -mattr=+v,+f -S| FileCheck %s
+; RUN: opt < %s --prefer-predicate-over-epilogue=predicate-dont-vectorize --passes=loop-vectorize -mcpu=sifive-p470 -mattr=+v,+f -force-tail-folding-style=data-with-evl -S| FileCheck %s --check-prefixes=EVL
+; COM: From issue #109468
```
What does COM stand for?

ElvisWang123 replied:
COM for commands.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2253,7 +2253,9 @@ InstructionCost VPWidenMemoryRecipe::computeCost(ElementCount VF,
   }
-  if (!Reverse)
+  // If the store value is a live-in scalar value which is uniform, we don't
+  // need to calculate the reverse cost.
```
At the moment, the revers will be done by to store recipe, which is why it is included here.

Could you elaborate how this fixes a difference in legacy and VPlan-based cost model?  AFAICT the patch also extends the legacy cost model to include logic to skip shuffle costs for invariant store operands.

fhahn replied:
Thanks for elaborating 

IIUC the issue is that the legacy cost model doesn't know about EVL at all, but at the VPlan level we may not need a mask due to using EVL instead.

To match the legacy behavior, wouldn't it be better to implement computeCost for the EVL memory recipes and always include the mask cost. With a TODO to make this more accurate once the legacy cost model is retired?

ElvisWang123 replied:
Thanks, that is a good idea!

I will implement the computeCost for Load/Store EVL recipes and always calculate the mask cost.
This implementation can fix the difference between the legacy cost model and the vplan-based cost model.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -5666,7 +5666,10 @@ LoopVectorizationCostModel::getConsecutiveMemOpCost(Instruction *I,
-  if (Reverse)
+  const StoreInst *SI = dyn_cast<StoreInst>(I);
+  bool IsLoopInvariantStoreValue =
+      SI && Legal->isInvariant(const_cast<StoreInst *>(SI)->getValueOperand());
```
isInvariant uses SCEV to determine loop-invariance, while isLiveIn only returns true for values defined outside the VPlan. This may introduce additional divergences, where the operand is invariant via SCEV but defined inside the loop

ElvisWang123 replied:
You are right, using this implementation may introduce extra divergences. 
I think the implementation in the legacy cost model is what we want but we cannot get the analysis in the VPlanRecipes.

Do you have other better methods to figure out if the value is loop invariant in the VPlanRecipes?

ElvisWang123 committed:
[VPlan] Implement VPWidenStoreEVLRecipe::computeCost().

Currently the EVL recipes transfer the tail masking to the EVL.
But in the legacy cost model, the mask exist and will calculate the
instruction cost of the mask.
To fix the difference between the VPlan-based cost model and the legacy
cost model, we always calculate the instruction cost of the mask in the EVL recipes.

Note that we should remove the mask cost in the EVL recipes when we
don't need to compare to the legacy cost model.

This patch also fix #109468.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2464,6 +2464,43 @@ void VPWidenStoreEVLRecipe::execute(VPTransformState &State) {
+  TTI::TargetCostKind CostKind = TTI::TCK_RecipThroughput;
+
+  if (!Consecutive) {
+    // TODO: Using the original IR may not be accurate.
+    // Currently, ARM will use the underlying IR to calculate gather/scatter
+    // instruction cost.
+    const Value *Ptr = getLoadStorePointerOperand(&Ingredient);
+    assert(!Reverse &&
+           "Inconsecutive memory access should not have the order.");
+    return Ctx.TTI.getAddressComputationCost(Ty) +
+           Ctx.TTI.getGatherScatterOpCost(Ingredient.getOpcode(), Ty, Ptr,
+                                          IsMasked, Alignment, CostKind,
+                                          &Ingredient);
```
```suggestion
  return VPWidenStoreRecipe::computeCost();
```

Use cost from base class if possible?

If so, sink variable assignments closer to use

ElvisWang123 replied:
Sure. Reuse the `VPWidenMemoryRecipe::computeCost()` when the load/store is not consecutive or masked.

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/vectorize-force-tail-with-evl-uniform-store.ll:
```diff
@@ -0,0 +1,25 @@
+; RUN: opt < %s --prefer-predicate-over-epilogue=predicate-dont-vectorize --passes=loop-vectorize -mcpu=sifive-p470 -mattr=+v,+f
+; RUN: opt < %s --prefer-predicate-over-epilogue=predicate-dont-vectorize --passes=loop-vectorize -mcpu=sifive-p470 -mattr=+v,+f -force-tail-folding-style=data-with-evl
```
Do we need both run lines?

Please also check the IR output

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/vectorize-force-tail-with-evl-uniform-store.ll:
```diff
@@ -0,0 +1,25 @@
+  %exitcond.not = icmp eq i64 %indvars.iv.next, 3
+  br i1 %exitcond.not, label %for.end16, label %for.body9
+
+for.end16:                                        ; preds = %for.body9
```
```suggestion
exit:  
```

ElvisWang123 replied:
Removed, thanks.

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/vectorize-force-tail-with-evl-uniform-store.ll:
```diff
@@ -0,0 +1,25 @@
+  br label %for.body9
+
+for.body9:                                        ; preds = %entry, %for.body9
+  %indvars.iv = phi i64 [ %spec.select, %entry ], [ %indvars.iv.next, %for.body9 ]
```
```suggestion
  %iv  = phi i64 [ %spec.select, %entry ], [ %indvars.iv.next, %for.body9 ]
```

ElvisWang123 replied:
Renamed, thanks.

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/vectorize-force-tail-with-evl-uniform-store.ll:
```diff
@@ -0,0 +1,25 @@
+; RUN: opt < %s --prefer-predicate-over-epilogue=predicate-dont-vectorize --passes=loop-vectorize -mcpu=sifive-p470 -mattr=+v,+f
+; RUN: opt < %s --prefer-predicate-over-epilogue=predicate-dont-vectorize --passes=loop-vectorize -mcpu=sifive-p470 -mattr=+v,+f -force-tail-folding-style=data-with-evl
+; Generated from issue #109468.
```
Could you add an explanation to the test. IIUC the important bit is that the store doesn't need a mask with EVL?

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/vectorize-force-tail-with-evl-uniform-store.ll:
```diff
@@ -0,0 +1,25 @@
+for.body9:                                        ; preds = %entry, %for.body9
+  %indvars.iv = phi i64 [ %spec.select, %entry ], [ %indvars.iv.next, %for.body9 ]
+  %1 = sub nuw nsw i64 1, %indvars.iv
+  %arrayidx13 = getelementptr [3 x i64], ptr %0, i64 0, i64 %1
```
```suggestion
  %arrayidx13 = getelementptr i64, ptr %0, i64 %1
```

ElvisWang123 replied:
Removed, thanks.

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/vectorize-force-tail-with-evl-uniform-store.ll:
```diff
@@ -0,0 +1,25 @@
+  %spec.select = select i1 %cmp1.peel, i64 2, i64 0
+  br label %for.body9
+
+for.body9:                                        ; preds = %entry, %for.body9
```
```suggestion
loop:
```

ElvisWang123 replied:
Removed, thanks.

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/vectorize-force-tail-with-evl-uniform-store.ll:
```diff
@@ -0,0 +1,25 @@
+target datalayout = "e-m:e-p:64:64-i64:64-i128:128-n32:64-S128"
+target triple = "riscv64-unknown-linux-gnu"
+
+define void @lshift_significand(i32 %n, ptr nocapture writeonly %0) local_unnamed_addr #0 {
```
```suggestion
define void @evl_store_cost(i32 %n, ptr nocapture writeonly %dst) {
```

ElvisWang123 replied:
Renamed and removed, thanks.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2464,6 +2464,43 @@ void VPWidenStoreEVLRecipe::execute(VPTransformState &State) {
+                                          &Ingredient);
+  }
+
+  InstructionCost Cost = 0;
+  // We need to use the getMaskedMemoryOpCost() instead of getMemoryOpCost()
+  // here because the EVL recipes using EVL to replace the tail mask. But in the
+  // legacy model, it will always calculate the cost of mask.
+  // TODO: Using getMemoryOpCost() instead of getMaskedMemoryOpCost when we
+  // don't need to care the legacy cost model.
+  Cost += Ctx.TTI.getMaskedMemoryOpCost(Ingredient.getOpcode(), Ty, Alignment,
```
```suggestion
  // We need to use the getMaskedMemoryOpCost() instead of getMemoryOpCost()
  // here because the EVL recipes using EVL to replace the tail mask. But in the
  // legacy model, it will always calculate the cost of mask.
  // TODO: Using getMemoryOpCost() instead of getMaskedMemoryOpCost when we
  // don't need to care the legacy cost model.
  InstructionCost Cost = Ctx.TTI.getMaskedMemoryOpCost(Ingredient.getOpcode(), Ty, Alignment,
```

ElvisWang123 replied:
Updated, thanks.

fhahn reviewed (commented):
Thanks for the update! Do we need a similar change for VPWidenLoadEVL?

ElvisWang123 committed:
Address comments

Reuse VPWidenMemoryRecipe::computeCost().
Implement VPWidenLoadEVLRecipe::computeCost().
Rename basic blocks in the testcase.

fhahn commented on llvm/test/Transforms/LoopVectorize/RISCV/vectorize-force-tail-with-evl-uniform-store.ll:
```diff
@@ -0,0 +1,144 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt < %s --prefer-predicate-over-epilogue=predicate-dont-vectorize --passes=loop-vectorize -mcpu=sifive-p470 -mattr=+v,+f -S | FileCheck %s --check-prefixes=CHECK
```
Drop non-EVL check line, as this test checks for an EVL only issue

ElvisWang123 replied:
Dropped, thanks.

ElvisWang123 committed:
Address comments

ElvisWang123 commented on llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:
```diff
@@ -2364,6 +2364,32 @@ void VPWidenLoadEVLRecipe::execute(VPTransformState &State) {
+                                                  VPCostContext &Ctx) const {
+  if (!Consecutive || IsMasked) {
+    return VPWidenMemoryRecipe::computeCost(VF, Ctx);
+  }
```
Removed, thanks.

fhahn reviewed (approved):
LGTM, thanks!
